federation/pkg/federation-controller/replicaset/planner/planner.go

```diff
@@ -87,15 +99,41 @@ func (p *Planner) Plan(replicasToDistribute int64, availableClusters []string) m
 		plan[preference.clusterName] = min
 	}
 
+	// This map contains information how many replicas were assigned to
+	// the cluster based only on the current replica count and
+	// rebalance=false preference. It will be later used in remaining replica
+	// distribution code.
+	preallocated := make(map[string]int64)
+
+	if p.preferences.Rebalance == false {
+		for _, preference := range preferences {
+			planned := plan[preference.clusterName]
+			count, hasSome := currentReplicaCount[preference.clusterName]
+			if hasSome && count > planned {
+				target := count
+				if preference.MaxReplicas != nil {
+					target = minInt64(*preference.MaxReplicas, target)
+				}
+				if capacity, hasCapacity := estimatedCapacity[preference.clusterName]; hasCapacity {
+					target = minInt64(capacity, target)
+				}
+				extra := minInt64(target-planned, remainingReplicas)
+				remainingReplicas -= extra
+				preallocated[preference.clusterName] = extra
+				plan[preference.clusterName] = extra + planned
+			}
+		}
+	}
+
 	modified := true
 
 	// It is possible single pass of the loop is not enough to distribue all replicas among clusters due
 	// to weight, max and rounding corner cases. In such case we iterate until either
 	// there is no replicas or no cluster gets any more replicas or the number
 	// of attempts is less than available cluster count. Every loop either distributes all remainingReplicas
 	// or maxes out at least one cluster.
-	// TODO: This algorithm is O(clusterCount^2). When needed use sweep-like algorithm for O(n log n).
-	for trial := 0; trial < len(availableClusters) && modified && remainingReplicas > 0; trial++ {
+	// TODO: This algorithm is O(clusterCount^3). Find something faster.
+	for trial := 0; trial < 2*len(availableClusters)*len(availableClusters) && modified && remainingReplicas > 0; trial++ {
```

Comment:
The comment above claims:
```
"Every loop either distributes all remainingReplicas or maxes out at least one cluster."
```

If this is still true, then your change doesn't make sense. Otherwise it is outdated.

Please fix.